### PR TITLE
[npu] support triangle attention for llama

### PR DESCRIPTION
**Pull request by oahzxl**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/kernel/npu/mha/triangle_attn.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        context_layer = npu_confusion_transpose(
+            context_layer, [0, 2, 1, 3], [*new_context_layer_shape], True
+        )
+        # =========================
+        # Context layer. [b, sq, hp]
+        # =========================
+        return context_layer
+
+
+if __name__ == "__main__":
```
Comment (ver217) [inline colossalai/kernel/npu/mha/triangle_attn.py:122]: remove

File: colossalai/kernel/npu/mha/triangle_attn.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            context_layer_tmp = torch.matmul(attention_probs, v_layer)
+            return context_layer_tmp
+
+        q, k, v = [rearrange(x, "b s h d -> b h s d") for x in (q, k, v)]
+        origin_attn_mask = origin_attn_mask.to(torch.bool)
+        #  input shape: [b, hn, sq, hd]
+        bsz, head_num, sequence_len, head_dim = k.shape
+        sparse_groups = sequence_len // block_size
+        # Determine whether blocks size can be divided by sequence_length
+        flag = sequence_len == block_size * sparse_groups
```
Comment (ver217) [inline colossalai/kernel/npu/mha/triangle_attn.py:74]: `flag` is meaningless. Can we use a meaningful name?

File: colossalai/kernel/npu/mha/triangle_attn.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import torch
+from einops import rearrange
+
+HAS_NPU_TRIANGLE_ATTENTION = False
+try:
+    from torch_npu import npu_scaled_masked_softmax
```
Comment (ver217) [inline colossalai/kernel/npu/mha/triangle_attn.py:22]: The condition is different from `get_attention_kernel()`

**Review by ver217 — CHANGES_REQUESTED** [review]

File: colossalai/kernel/npu/mha/triangle_attn.py (hunk 3 above)
Comment (oahzxl) [reply colossalai/kernel/npu/mha/triangle_attn.py:22]: NPUColoAttention will use spda attention by default. If it has npu_scaled_masked_softmax kernel, we will use triangle attention instrad.

File: colossalai/kernel/npu/mha/triangle_attn.py (hunk 2 above)
Comment (oahzxl) [reply colossalai/kernel/npu/mha/triangle_attn.py]: ok, rename to divisible_flag

File: colossalai/kernel/npu/mha/triangle_attn.py (hunk 1 above)
Comment (oahzxl) [reply colossalai/kernel/npu/mha/triangle_attn.py]: ok, removed now

File: colossalai/kernel/npu/mha/mha.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+from .sdpa_attn import npu_sdpa_attention
+from .triangle_attn import HAS_NPU_TRIANGLE_ATTENTION
+
+
+class NPUColoAttention(torch.nn.Module):
+    def __init__(self, embed_dim: int, num_heads: int, dropout: float = 0.0, scale: float = None):
+        super().__init__()
+
+        try:
```
Comment (ver217) [inline colossalai/kernel/npu/mha/mha.py:14]: Is this removed be pre-commit? Try `import torch_npu # noqa`

File: colossalai/kernel/npu/mha/sdpa_attn.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
+import torch
+from einops import rearrange
+
+import torch_npu  # noqa
```
Comment (ver217) [inline colossalai/kernel/npu/mha/sdpa_attn.py:4]: It seems that this is not used. Why import it?

**Review by ver217 — CHANGES_REQUESTED** [review]

**Review by ver217 — APPROVED** [review]

**Pull request merged**